robot-server/robot_server/service/task_runner.py

```diff
@@ -36,15 +34,45 @@ def run(self, func: TaskFunc, **kwargs: Any) -> None:
 
         Arguments:
             func: An async, None-returning function to run in the background.
-            Use kwargs to add arguments if required.
+            Use **kwargs to pass to func.
         """
         func_name = func.__qualname__
 
-        async def _run_async_task() -> None:
+        async def wrapper() -> None:
+            # Add the previous logic back
+            current_task = asyncio.current_task()
             try:
                 await func(**kwargs)
                 log.debug(f"Background task {func_name} succeeded")
             except Exception as e:
                 log.warning(f"Background task {func_name} failed", exc_info=e)
+            finally:
+                self._running_tasks.remove(current_task)
+
+        wrapper_task = asyncio.create_task(wrapper())
+        self._running_tasks.add(wrapper_task)
+
+    async def cancel_all_and_clean_up(self) -> None:
+        for task in self._running_tasks:
+            task.cancel()
+        await asyncio.gather(*self._running_tasks, return_exceptions=True)
+
+
+def initialize_task_runner(app_state: AppState) -> None:
+    """Create a new `TaskRunner` and store it on `app_state`."""
+    _init_taskrunner.set_on(app_state, TaskRunner())
+
+
+async def clean_up_task_runner(app_state: AppState) -> None:
+    """Clean up the `TaskRunner` stored on `app_state`."""
```

Comment:
Docstring suggestion to help prevent misuse:

```suggestion
async def clean_up_task_runner(app_state: AppState) -> None:
    """Clean up the `TaskRunner` stored on `app_state`.
    
    Intended to be called just once, when the server shuts down.
    """
```